src/ggml.c

```diff
@@ -7758,11 +7652,14 @@ struct ggml_tensor * ggml_conv_2d(
     int                  d0,
     int                  d1) {
 
-    struct ggml_tensor * result = ggml_conv_2d_stage_0(ctx, a, b, s0, s1, p0, p1, d0, d1); // [N, OH, OW, IC * KH * KW]
-    result = ggml_conv_2d_stage_1(ctx, a, result);
-
+    struct ggml_tensor * result = ggml_im2col(ctx, a, b, s0, s1, p0, p1, d0, d1, true); // [N, OH, OW, IC * KH * KW]
+    result = ggml_reshape_4d(ctx,
+            ggml_cont(ctx, ggml_transpose(ctx,
+            ggml_mul_mat(ctx,
+            ggml_cont(ctx, ggml_reshape_2d(ctx, a, (a->ne[0] * a->ne[1] * a->ne[2]),  a->ne[3])),
+            ggml_cont(ctx, ggml_reshape_2d(ctx, result, result->ne[0],  (result->ne[3] * result->ne[2] * result->ne[1])))))),
+            result->ne[1], result->ne[2], a->ne[3], result->ne[3]); // [N, OC, OH, OW] = [OC, IC * KH * KW] x [N*OH*OW, IC * KH * KW]
```

Comment:
Do we actually need all these `ggml_cont`?

If a tensor is already contiguous, after `ggml_reshape` you don't need to use `ggml_cont`.
This should eliminate the inner 2 `ggml_cont`

The outer one seems unnecessary - we just need to declare that the `ggml_conv` op returns a transposed result and it will be up to the user to transpose it if they need to.